---
title: "R Notebook"
output: 
  html_notebook:
    toc: TRUE
    toc_depth: 2
    toc_float: TRUE
---

# 1. libraries and palette
```{r, results=F}
suppressPackageStartupMessages({
  library(Seurat)
  library(dplyr)
  library(Signac)
  library(ggplot2)
  library(ggsci)
  library(patchwork)
  library(GenomicRanges)
  library(AnnotationHub)
  library(SeuratWrappers)
  library(cicero)
  library(JASPAR2020)
  library(patchwork)
  library(TFBSTools)
  library(motifmatchr)
  library(BSgenome.Drerio.UCSC.danRer11)
  library(dittoSeq)
}) 
options(future.globals.maxSize = 4000 * 1024^2)
options(Seurat.object.assay.version = "v3")
```

```{r}
mypal <- pal_igv(palette = "default",alpha = 1)(40)
```

# 2. Read data
```{r}
rhom <- readRDS(file = "RDSfiles/HB1.r2.r3.RDS")
```

## 2.1 list of zebrafish TFs
This was generated from list of human TF from http://humantfs.ccbr.utoronto.ca/index.php and zfin zebrafish/human ortholog table in R notebook, R_General/GetListZebrafishTFs.Rmd
```{r}
z.tf <- read.csv(file = "~/Documents/Projects/R_General/zfin_ortho_humanTFs.txt", header = T, sep = "\t")
head(z.tf)
```

# 3. DE genes over time
## 3.1 HB.1 (E10) vs r2 (HB13)
```{r, results=F}
DefaultAssay(rhom) <- "RNA"
Idents(rhom) <- "Clusters"
HB1vsr2.markers <- FindMarkers(rhom, ident.1 = "HB.1", ident.2 = "r2")
HB1vsr2.markers
```

### 3.1.1 DE genes p-value < 0.05, |log2FC| > 1.5
```{r}
HB1vsr2.filtered <-  HB1vsr2.markers[HB1vsr2.markers$p_val_adj < 0.05 & abs(HB1vsr2.markers$avg_log2FC) > 1.5,]
HB1vsr2.filtered
```

## 3.2 HB.1 vs r3
```{r, results=F}
DefaultAssay(rhom) <- "RNA"
Idents(rhom) <- "Clusters"
HB1vsr3.markers <- FindMarkers(rhom, ident.1 = "HB.1", ident.2 = "r3")
HB1vsr3.markers
```

### 3.2.1 DE genes p-value < 0.05, |log2FC| > 1.5
```{r}
HB1vsr3.filtered <-  HB1vsr3.markers[HB1vsr3.markers$p_val_adj < 0.05 & abs(HB1vsr3.markers$avg_log2FC) > 1.5,]
HB1vsr3.filtered
```

## 3.3 heatmap

* DE genes between HB.1 vs r3 (adj p-value < 0.05 and |log2FC| > 1.5) not in HB.1 vs r2 (only adj p-value < 0.05) 
* plus DE genes between HB.1 and r2 (adj p-value < 0.05 and |log2FC| > 1.5) not in HB.1 and r3 (only adj p-value < 0.05)

```{r}
HB1.r3.unique2 <- rownames(HB1vsr3.filtered)[!rownames(HB1vsr3.filtered) %in% 
                                               rownames(HB1vsr2.markers[HB1vsr2.markers$p_val_adj < 0.05,])]
HB1.r2.unique2 <- rownames(HB1vsr2.filtered)[!rownames(HB1vsr2.filtered) %in% 
                                               rownames(HB1vsr3.markers[HB1vsr3.markers$p_val_adj < 0.05,])]
```

```{r, fig.width=7, fig.height=10}
DefaultAssay(rhom) <- "RNA"
comb_heatmap_all <- dittoHeatmap(rhom, 
                             genes = c(HB1.r3.unique2,HB1.r2.unique2), 
                             #cells = cells.rhb, 
                             assay = "RNA", #slot = "data", 
                             annot.by = c("orig.ident","Clusters"), 
                             order.by = c("orig.ident","Clusters"), 
                             scaled.to.max = T, 
                             show_rownames = T,
                             cluster_rows = T)
```

```{r}
png(filename = "../results/SuppFig6_panelC_heatmap_HB1_r2_r3.png", width=7,height=10,units="in",res=1200)
DefaultAssay(rhom) <- "RNA"
comb_heatmap_all <- dittoHeatmap(rhom, 
                             genes = c(HB1.r3.unique2,HB1.r2.unique2), 
                             #cells = cells.rhb, 
                             assay = "RNA", #slot = "data", 
                             annot.by = c("orig.ident","Clusters"), 
                             order.by = c("orig.ident","Clusters"), 
                             scaled.to.max = T, 
                             show_rownames = T,
                             cluster_rows = T)
dev.off()
```

```{r}
sessionInfo()
```

